### PR TITLE
V4.0 scratch and flush

### PR DESCRIPTION
**Pull request by aviviadi**

### CODE REVIEW

File: src/Voron/Impl/Scratch/ScratchBufferPool.cs [hunk 1]
```diff
@@ -183,7 +183,14 @@ public PageFromScratchBuffer Allocate(LowLevelTransaction tx, int numberOfPages)
 
                 bool createNextFile = false;
 
-                if (current.File.HasDiscontinuousSpaceFor(tx, size, _scratchBuffers.Count))
+                if (tx.IsLazyTransaction)
```
Comment (aviviadi) [inline src/Voron/Impl/Scratch/ScratchBufferPool.cs:186]: this is the easiest solution that works (even if not super efficient in terms of reusing scratch, it keeps us supporting any size and rate of bulk insertions using lazy commits and keeping scratch size on ~500MB).

File: src/Raven.Server/Documents/BulkInsert/BulkInsertConnection.cs [hunk 2]
```diff
@@ -255,7 +258,30 @@ private void FlushDocuments(DocumentsOperationContext context, List<BulkInsertDo
                             // flush everything
                             tx.Commit();
                         }
-                        context.Environment().ForceLogFlushToDataFile(null, true);
+                        bool lockTaken = false;
+                        var journal = _database.DocumentsStorage.Environment.Journal;
+                        Debug.Assert(journal != null);
+                        Console.WriteLine("Going to tryTakeLock. retry=" + retry);
```
Comment (ayende) [inline src/Raven.Server/Documents/BulkInsert/BulkInsertConnection.cs:264]: Need to remove this

File: src/Raven.Server/Documents/BulkInsert/BulkInsertConnection.cs [hunk 3]
```diff
@@ -255,7 +258,30 @@ private void FlushDocuments(DocumentsOperationContext context, List<BulkInsertDo
...
+                        {
+                            Console.WriteLine("TryTakeFlushingLock = " + lockTaken);
+                            if (lockTaken == false)
+                            {
+                                var now = SystemTime.UtcNow;
+
+                                // lets wait for the flush to end and retry put doc
+                                using (journal.Applicator.TryTakeFlushingLock(ref lockTaken, TimeSpan.FromSeconds(30)))
+                                {
+                                    Console.WriteLine("exiting = " + lockTaken + " | after " +
```
Comment (ayende) [inline src/Raven.Server/Documents/BulkInsert/BulkInsertConnection.cs:275]: Replace with logs

File: src/Raven.Server/Documents/BulkInsert/BulkInsertConnection.cs [hunk 4]
```diff
@@ -255,7 +258,30 @@ private void FlushDocuments(DocumentsOperationContext context, List<BulkInsertDo
...
                             tx.Commit();
                         }
-                        context.Environment().ForceLogFlushToDataFile(null, true);
+                        bool lockTaken = false;
+                        var journal = _database.DocumentsStorage.Environment.Journal;
+                        Debug.Assert(journal != null);
+                        Console.WriteLine("Going to tryTakeLock. retry=" + retry);
+                        using (journal.Applicator.TryTakeFlushingLock(ref lockTaken))
+                        {
+                            Console.WriteLine("TryTakeFlushingLock = " + lockTaken);
```
Comment (ayende) [inline src/Raven.Server/Documents/BulkInsert/BulkInsertConnection.cs:267]: Need to remove it

**Pull request merged**